src/coprocessor/dag/rpn_expr/function.rs

```diff
@@ -32,6 +49,267 @@ impl<T: RpnFunction + ?Sized> RpnFunction for Box<T> {
     fn args_len(&self) -> usize {
         (**self).args_len()
     }
+
+    #[inline]
+    fn eval(
+        &self,
+        rows: usize,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue> {
+        (**self).eval(rows, context, payload)
+    }
+}
+
+pub struct Helper;
+
+impl Helper {
+    /// Evaluates a function without argument to produce a vector value.
+    ///
+    /// The function will be called multiple times to fill the vector.
+    #[inline(always)]
+    pub fn eval_0_arg<Ret, F>(
+        rows: usize,
+        mut f: F,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue>
+    where
+        Ret: Evaluable,
+        F: FnMut(&mut EvalContext, RpnFnCallPayload<'_>) -> Result<Ret>,
+    {
+        assert_eq!(payload.args_len(), 0);
+
+        let mut result = Vec::with_capacity(rows);
+        for _ in 0..rows {
+            result.push(f(context, payload)?);
+        }
+        Ok(Ret::into_vector_value(result))
+    }
+
+    /// Evaluates a function with 1 scalar or vector argument to produce a vector value.
+    ///
+    /// The function will be called multiple times to fill the vector.
+    #[inline(always)]
+    pub fn eval_1_arg<Arg0, Ret, F>(
+        rows: usize,
+        mut f: F,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue>
+    where
+        Arg0: Evaluable,
+        Ret: Evaluable,
+        F: FnMut(&mut EvalContext, RpnFnCallPayload<'_>, &Arg0) -> Result<Ret>,
+    {
+        assert_eq!(payload.args_len(), 1);
+
+        let mut result = Vec::with_capacity(rows);
+        if payload.raw_arg_at(0).is_scalar() {
+            let v = Arg0::borrow_scalar_value(payload.raw_arg_at(0).scalar_value().unwrap());
+            for _ in 0..rows {
+                result.push(f(context, payload, v)?);
+            }
+        } else {
+            let v = Arg0::borrow_vector_value(payload.raw_arg_at(0).vector_value().unwrap());
+            assert_eq!(rows, v.len());
+            for i in 0..rows {
+                result.push(f(context, payload, &v[i])?);
+            }
+        }
+        Ok(Ret::into_vector_value(result))
+    }
+
+    /// Evaluates a function with 2 scalar or vector arguments to produce a vector value.
+    ///
+    /// The function will be called multiple times to fill the vector.
+    #[inline(always)]
+    pub fn eval_2_args<Arg0, Arg1, Ret, F>(
+        rows: usize,
+        f: F,
+        context: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue>
+    where
+        Arg0: Evaluable,
+        Arg1: Evaluable,
+        Ret: Evaluable,
+        F: FnMut(&mut EvalContext, RpnFnCallPayload<'_>, &Arg0, &Arg1) -> Result<Ret>,
+    {
+        assert_eq!(payload.args_len(), 2);
+
+        if payload.raw_arg_at(0).is_scalar() {
```

Comment:
How about using variables for these repeated function calls, like `payload.raw_arg_at()`?

Reply:
raw_arg_at has a inline, which will finally becomes an array index access, so it should be fine.